### PR TITLE
Fix recurring payments on Indian CCs

### LINKED ISSUE #2935: Update 3DS flow for recurring transactions on Indian CCs
### Description
New regulations for processing recurring transactions in India are going live on September 30.
Changes to the 3DS flow are needed.

### Acceptance criteria

* Transactions on India issued CCs use correct flow and APIs.
* Transactions work in shortcode checkout.
* Transactions work in UPE checkout.
* Transactions work in blocks checkout.

### Designs
N/A

### Testing instructions

1. Purchase subscription product.
2. Trigger renewal.
3. Check new flow is applied.

### Dev notes

Not available yet.

### Additional context

Internal discussion and context: pc4etw-ky-p2

### PR DESCRIPTION
**Pull request by ismaeldcom**
Fixes #2935

### Changes proposed in this Pull Request
With the new regulations to process recurring payments for cards issued by Indian banks, we need to generate and save a mandate ID, which will be attached to every renewal order.

- Generate mandate params from subscriptions details. Implement `get_mandate_params_for_order` and call it within `process_payment` and `process_payment_for_order` using `$additional_api_parameters`.
- Save the returned mandate ID from payment intent succeed webhook on the order meta as `_stripe_mandate_id`.
- Attach mandate to renewal orders if required. Implement `get_mandate_param_for_renewal_order` and call it within `scheduled_subscription_payment` using `$additional_api_parameters`.
- Add an order note about the required authorization of the payment by the merchant if required. Implement `maybe_add_customer_notification_note` and call it within `process_payment_for_order` with the new `$processing` property.
- Add `$processing` property to `WC_Payments_API_Intention` to be able to access `customer_notification`.
- Update intent failed webhook handler to support intents without charges, using `last_payment_error` instead. Add messages for the new codes, and include the Stripe error message with our default one.

### Known issues
- Changing payment method on a subscription which have a mandate will result in a failed renovation. Using another Indian CC will result in an error by mandate not matching the payment method. And using any other card will also fail because we will attach the initial mandate.
  - **Possible fix**: Update the mandate on payment method change, removing it if it's empty.
- Mandates are created in a pending state and transition to active after 60-90 seconds. If we attempt to renew a subscription within that time frame, it will fail and add an order note with the following message: `Error: Only active mandates can be used with PaymentIntents.`
  - **Possible fix**: Delay the subscription renovation until the mandate is ready to be used, listening to `mandate.updated` webhook.

### Testing instructions

> **Note**: Stripe simulates the async nature of the renewal process delaying 5-10 minutes both success and failure responses.

#### Requirements
- Place orders in `INR` using Multi-currency or updating your store currency.
- Be sure you're listening to Stripe webhooks as the implementation relies on them. (`npm run listen`)
- Create a subscription product below `₹15,000 INR` and another one over it.

#### WC Subscriptions – Single subscription
- Enable WooCommerce Subscriptions and be sure you have early renewals enabled.
- Place and order with the following cards and subscription amounts.
  - 4000003560000123 – **less than** ₹15k
  - 4000003560000123 – **greater than** ₹15k
  - 4000003560000297 – any amount
  - 4000003560000248 – any amount
  - 4000003560000263 – any amount
- Wait a minute for the mandate to be active.
- Go to **WooCommerce → Status → Scheduled Actions → Pending** and filter by `woocommerce_scheduled_subscription_payment`.
- Run all the hooks where `subscription_id` matches the 5 subscriptions you have just created.
- Wait patiently for Stripe webhook 🕙.
- Order notes should match the following table.

| *123 <15k | *123 >15k | *297 | *248 | *263 |
| - | - | - | - | - |
|success|success-over-15k|fail-cancel-pause-297|fail-notification-248|fail-mandate-cancelled-263|

#### WC Subscriptions – Multiple subscriptions
- Enable WooCommerce Subscriptions and be sure you have early renewals enabled.
- Place an order using card `4000003560000123` with multiple subscriptions with:
  - Same interval.
  - Different interval.
- Wait a minute for the mandate to be active.
- Go to **WooCommerce → Status → Scheduled Actions → Pending** and filter by `woocommerce_scheduled_subscription_payment`.
- Run all the hooks where `subscription_id` matches the 2 subscriptions you have just created.
- Wait patiently for Stripe webhook 🕙.
- Renewals should succeed.

#### UPE vs classic checkout
The previous testing instructions should work for both UPE and classic checkout. Please test it using the missing one.

#### WCPay Subscriptions
Does not apply.

#### Known issues
| Pending mandate | Wrong mandate | Not matching mandate |
| - | - | - |
|pending-mandate|wrong-mandate|not-matching-mandate|

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**
- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : [Testing instructions](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-5.2.0#fix-recurring-payments-on-indian-ccs--pr3011)
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (ismaeldcom) [inline includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:725]: Appreciate any input here. I think we should apply it to every order with subscriptions, and probably not here.

Comment (ismaeldcom) [inline includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:725]: Appreciate any input here. I think we should apply it to every order with subscriptions, and probably not here.

File: includes/admin/class-wc-rest-payments-webhook-controller.php
Comment (ismaeldcom) [inline includes/admin/class-wc-rest-payments-webhook-controller.php:268]: This is quite deep! I opted for the null coalescing operator instead of checking for the existence of each key. Apart from that, not sure if there is a better place to put it. The mandate id is only included in the payment_intent.succeed so we need to rely on the event to get it.

**Comment by ismaeldcom** [comment]
I'm wondering if we should handle the payment intent failed response from Stripe to add another order note and provide some more details on why it failed, to the merchant.

For now, Stripe sent us 2 cards that fail with a specific decline code (mentioned in testing instructions). We could handle those 2, but to do that we will need to also update the server to forward the `payment_intent.payment_failed` to the client. AFAIK it isn't being forwarded for now.

I'll appreciate any thought here. I see it as a bit out of the scope of this PR. Should we implement it here? Create another issue? Or is there a reason to not do this?

cc @Automattic/sigma

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (dmallory42) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:739]: I agree it makes sense to apply it to every Subscription order. As per the Stripe docs:

> You can still accept off-session payments with a card set up for on-session payments, but the bank is more likely to reject the off-session payment and require authentication from the cardholder.

It seems to me that since subscriptions will always have renewal payments, `off_session` makes the most sense here. Maybe we can attach it in a different place, such as when we set up the payment intent? I'm not really sure where the best place for this is.

File: includes/admin/class-wc-rest-payments-webhook-controller.php
Comment (dmallory42) [reply includes/admin/class-wc-rest-payments-webhook-controller.php:277]: Even though it's very deep, I can't really think of a better way to do this.

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (ismaeldcom) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:739]: Thanks! I'll need to think more about where to put it. For now, we only set this in `class-wc-payments-api-client.php` if the payment method will be saved. But it isn't enough for this use case. Cause `setup_future_usage=off_session` is always required with e-mandates, and if we use an already saved card will not work.

**Comment by daquinons** [comment]
> Another question about order notes involving greater changes and a bit out of the scope in this PR.
> 
> If the mandate becomes invalid the merchant will see the following error. Looks like we just append the error message from Stripe on subscription payment fails. And not sure if we should improve at least this message to provide more useful information. Something like `Payment failed due to a wrong mandate, you will need to ask the customer for a new on session payment...`
> 
> What do you all think? Should we create another issue for this?
> 
> Screenshot 2021-10-07 at 11 11 58

Yes, definitely this deserves an issue. From the places I've seen we add notes into the order, they are always very user friendly to add context to the merchant, so good idea to create a new issue! 👍

**Review by dmallory42 — APPROVED** [review]
Tested and all working perfectly after we finished debugging, nice work! 👍 I think this is ready to 🚢  once tests are done.

My thoughts on the comments you raised are as follows:

> I'm wondering if we should handle the payment intent failed response from Stripe to add another order note and provide some more details on why it failed, to the merchant.

> For now, Stripe sent us 2 cards that fail with a specific decline code (mentioned in testing instructions). We could handle those 2, but to do that we will need to also update the server to forward the payment_intent.payment_failed to the client. AFAIK it isn't being forwarded for now.

> I'll appreciate any thought here. I see it as a bit out of the scope of this PR. Should we implement it here? Create another issue? Or is there a reason to not do this?

I think it would be great for us to provide a generalised order note for example "A request to authorise this payment failed with the following error code: `transaction_not_approved`", but I think given we want to get this fix out, it would be fine to do this in a follow up issue.

I also agree with David about handling the other error message an a more friendly way but again this can be done in a followup issue IMO.

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (reykjalin) [inline includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:731]: I think you may need to add some special handling for orders where the initial payment doesn't have the same value as any of the renewal payments 🤔 

I think I handled this correctly in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981/commits/864cbb8a40dd4f76dc4c6feea40cad83f8e64f05, but I wasn't able to test it [due to API issues](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981#issuecomment-939246499).

Kudos to @ricardo for spotting this problem in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981#issuecomment-939208232

**Review by reykjalin — CHANGES_REQUESTED** [review]
This looks good to me! I don't see any differences between this PR and my own PR in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1981.

@ricardo found an issue in my PR that I think you have to address in this PR as well, but otherwise this is looking good 👌

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (ismaeldcom) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:731]: Thanks @ricardo and @reykjalin for pointing it out! I missed it too.

Addressed in 2e53900

**Review by htdat — COMMENTED** [review]
Nitpick: @ismaeldcom - can you help add changelog entries for readme.txt and changelog.txt?

**Review by reykjalin — APPROVED** [review]
LGTM 🙂

**Comment by ismaeldcom** [comment]
Found an issue with UPE testing it again today. The mandate options were not added to subscriptions using UPE and a new card. Because we use another process, creating the intent without confirm, where I initially added the new code.

For it to work we need to add the `payment_method_options` param to `update_intention` on server 1241-gh-Automattic/woocommerce-payments-server.

**To ensure this is properly covered, I'll appreciate a pair of hands testing it. Will be enough to just place a subscription order following the main testing instructions, but using a new card, you just need to remove it if they are already used.**

Until #3118 got merged, would be better to add something like `sleep( 10 );` to `process_webhook_payment_intent_succeeded` if you experience some issues with the order not being found on `payment_intent.succeeded`.

**Review by dmallory42 — APPROVED** [review]
Tested the updated PR and all working as expected. Nice job! :shipit:

**Comment by haszari** [comment]
Tagging @Automattic/helix for a review as requested. Note I've removed the `pr: ready to merge` label as it conflicts with the `on hold` status and to avoid confusion or accidental merge.

**Comment by aprea** [comment]
@haszari I don't think this PR is ready for review yet given the merge conflicts. It looks like there is a plan to resolve these and potentially tweak some other code:

>The PR is now a bit outdated, we will need to solve any conflict, update the code and double-check that everything works as expected.

>Update PR code, test all scenarios, and ensure it’s compatible with WC Pay Subscriptions.

Probably best to wait until these are completed before we review & test this.

**Comment by daquinons** [comment]
> @haszari I don't think this PR is ready for review yet given the merge conflicts. It looks like there is a plan to resolve these and potentially tweak some other code:
> 
> > The PR is now a bit outdated, we will need to solve any conflict, update the code and double-check that everything works as expected.
> 
> > Update PR code, test all scenarios, and ensure it’s compatible with WC Pay Subscriptions.
> 
> Probably best to wait until these are completed before we review & test this.

Yep, thank you @haszari, we appreciate the help! 

@ismaeldcom will work on putting the PR on shape during the week and he'll ping you when it's ready for review 👍

**Comment by haszari** [comment]
> @ismaeldcom will work on putting the PR on shape during the week and he'll ping you when it's ready for review 👍

@ismaeldcom feel free to request a review from `@woocommerce/helix` any time, that should get it in our reviewer queue.

**Comment by ismaeldcom** [comment]
Hey @Automattic/helix! We finally have the PR ready for review. I've made some updates to the code as a lot of things have changed since the initial implementation.

I would like to know if we need to consider anything else for WCPay Subscriptions and if there is a way to test manual renewals like in WC Subs.

Thanks!

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (oaratovskyi) [inline includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php]: nitpick: we can return early here `if ( ! $this->is_subscriptions_enabled() )`, but this implementation is also good!

File: includes/class-wc-payments-webhook-processing-service.php
Comment (oaratovskyi) [inline includes/class-wc-payments-webhook-processing-service.php:698]: I wonder if we need here the `default:` case with `The payment was not able to be processed` failure message? I think it might be helpful if Stripe adds another code that we are not aware of

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (oaratovskyi) [inline includes/wc-payment-api/class-wc-payments-api-client.php]: nitpick: I see that doc comments are indented in three columns above, this breaks the rule 😅  But again, it's not important, the linter is not complaining

**Review by oaratovskyi — APPROVED** [review]
@ismaeldcom first of all - big thanks for working on this one! I know it wasn't an easy one and it was blocked a couple of times, great to see it's ready now 🎉 

I followed testing instructions very accurately (classic + UPE). I can confirm that it works as expected: successful and failed renewals for different reasons. I also noticed that after successful renewal another one is scheduled depending on the subscription frequency. 

The code is clean and straightforward, documented very well, and covered by tests, I only described a few nitpicks, overall the code is looking great 💪 

I know that it's still open whether it needs additional work regarding WCPay Subscriptions, but I am pre-approving the PR and looking forward to feedback from the Helix team.

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (ismaeldcom) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php]: Good catch! I missed it, I always prefer early returns too 🙂  Addressed in d220c6d

File: includes/class-wc-payments-webhook-processing-service.php
Comment (ismaeldcom) [reply includes/class-wc-payments-webhook-processing-service.php:698]: It's actually covered with [this new return](https://github.com/Automattic/woocommerce-payments/blob/d26264a205f503b5691c07e263302117c7f35380/includes/class-wc-payments-webhook-processing-service.php#L667). And instead of returning just `The payment was not able to be processed` I also added Stripe's error message.

Maybe it's a bit confusing because I rewrote the switch to use `return` rather than `break`. So using a single return at the end will act as the `default`. I prefer it when we only want to return something, and especially if we have a nested `switch`.

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (ismaeldcom) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: Thanks for noticing it, I would like it to be fixed by PHPCBF! Addressed in d220c6d

File: includes/class-wc-payments-webhook-processing-service.php
Comment (oaratovskyi) [reply includes/class-wc-payments-webhook-processing-service.php:698]: Aaaah, agree, makes sense.

**Comment by haszari** [comment]
FYI @shendy-a8c  - this needs a Helix review, maybe already on your radar?

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (shendy-a8c) [inline includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:900]: In `get_mandate_params_for_order()` you check if `is_subscriptions_enabled()`, I think it's a good idea to also check here in `get_mandate_param_for_renewal_order()` as well.

Very likely `is_subscriptions_enabled()` returns true because `woocommerce/subscriptions-core` is [packaged](https://github.com/Automattic/woocommerce-payments/blob/c6154cddf73f49e486ee9316da1075c9cf982418/composer.json#L32) in WCPay but just to be safe because otherwise calling `wcs_get_subscriptions_for_renewal_order()` will return a fatal error.

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (shendy-a8c) [inline includes/wc-payment-api/class-wc-payments-api-client.php]: I don't see any `create_intention()` calls that is passed by `$additional_parameters`. Do I miss anything?

File: includes/class-wc-payments-webhook-processing-service.php
Comment (shendy-a8c) [inline includes/class-wc-payments-webhook-processing-service.php:380]: Same with my above comment. I don't understand this change.

I see that originally the value of `$payment_method_id` is from `data.object.charges.data[0].payment_method` and now it's from `data.object.last_payment_error.id`.

File: includes/class-wc-payments-webhook-processing-service.php
Comment (shendy-a8c) [inline includes/class-wc-payments-webhook-processing-service.php:366]: I see that originally the value of `$payment_method_type` is from `data.object.charges.data[0].payment_method_details.type` and now it's from `data.object.last_payment_error.type`.

I don't understand why that change. Will we always get `data.object.last_payment_error` for `payment_intent.payment_failed` webhook event despite the failing payment method (`card`, `us_bank_account`, or `au_becs_debit`)?

Will this change work when payment intent failed for `us_bank_account` and `au_becs_debit`? I don't know how to test that. Do you know?

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (shendy-a8c) [inline includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:900]: Is this intended that the function name is `get_mandate_param_for_renewal_order` instead of `get_mandate_params_for_renewal_order`?

Just a `nit` because it is not symmetrical with `get_mandate_params_for_order`.

**Review by shendy-a8c — COMMENTED** [review]
Is it intended to always pass mandate params (from `get_mandate_params_for_order()` and `get_mandate_param_for_renewal_order()`) even though payment is using a card not issued by India bank?

I was expecting a check if card is issued by India bank and only then pass the mandate parameters but maybe checking which bank country issues the card isn't possible so we have to always pass mandate params to Stripe and Stripe will return mandate id when we receive the `payment_intent.succeeded` webhook event.

**Comment by shendy-a8c** [comment]
Is mandate tied to specific card? If someone knows a mandate id, can they use it to create a payment intent with different card and by pass the needed customer confirmation?

**Comment by shendy-a8c** [comment]
> I would like to know if we need to consider anything else for WCPay Subscriptions and if there is a way to test manual renewals like in WC Subs.

Instruction to test WCPay Subscription renewal can be found here: PCYsg-DQ4-p2.

**Comment by shendy-a8c** [comment]
I saw this PR branch was pretty behind `develop`, so I merged develop into `fix/2935-recurring-transactions-indian-ccs`. I hope it's OK.

**Comment by shendy-a8c** [comment]
> Known issues
Changing payment method on a subscription which have a mandate will result in a failed renovation. Using another Indian CC will result in an error by mandate not matching the payment method. And using any other card will also fail because we will attach the initial mandate.
Possible fix: Update the mandate on payment method change, removing it if it's empty.

Will this be addressed, maybe in different PR? If so, please create a separate issue so it is not forgotten.

Changing payment method is part of the critical flows ([ref](https://github.com/Automattic/woocommerce-payments/wiki/WC-Pay-Subscriptions-critical-flows-testing-instructions#change-payment-method-card-for-an-active-subscription)) but I bet it won't be detected as bug because GlobalStep likely will test using US card.

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (ismaeldcom) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:900]: I wonder if it's necessary when we will only call `get_mandate_param_for_renewal_order()` within `scheduled_subscription_payment` 🤔 I assume it could only be called if Subscriptions are enabled.

In the other hand `get_mandate_params_for_order()` could be called for an order that's not a subscription.

WDYT?

File: includes/wc-payment-api/class-wc-payments-api-client.php
Comment (ismaeldcom) [reply includes/wc-payment-api/class-wc-payments-api-client.php]: Good catch! It was used in a previous implementation and I forgot to revert it. Addressed in 2777b24

File: includes/class-wc-payments-webhook-processing-service.php
Comment (ismaeldcom) [reply includes/class-wc-payments-webhook-processing-service.php:366]: Reply to this and:
> I see that originally the value of $payment_method_id is from data.object.charges.data[0].payment_method and now it's from data.object.last_payment_error.id.

I mentioned it on the PR description, but I'll give more context here 🙂 

There are some `payment_intent.payment_failed` that doesn't have any charges. With our current implementation that will result in an on hold order with no additional notes. Instead, we can rely on `data.object.last_payment_error` which contains the error reason, among other things.

There is no existing issue, but we've seen similar cases – a recent Slack conversation about it p1666137895802399/1665973745.206889-slack-CGGCLBN58

It will also work for `us_bank_account` and `au_becs_debit`. As an example [here is a BECS failed intent](https://dashboard.stripe.com/test/connect/accounts/acct_1M4juUQoftj2Jyyc/payments/pi_3M4jyVQoftj2Jyyc17kpQBwj).

To test BECS you'll need to onboard an AU Stripe account, and manually request BECS capability through Stripe's dashboard before being able to enable it in WCPay Settings.

My original idea was to let this part out of the PR and move it to a follow-up. But I think having failed orders on-hold is not an acceptable tradeoff.

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (ismaeldcom) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:900]: Yep, I've used the word in singular here because we're only adding a single parameter. Do you think it will sound better if we use the plural in both of them?

**Comment by ismaeldcom** [comment]
Thanks for the detailed review @shendy-a8c! 

>Is it intended to always pass mandate params (from get_mandate_params_for_order() and get_mandate_param_for_renewal_order()) even though payment is using a card not issued by India bank?
>
>I was expecting a check if card is issued by India bank and only then pass the mandate parameters but maybe checking which bank country issues the card isn't possible so we have to always pass mandate params to Stripe and Stripe will return mandate id when we receive the payment_intent.succeeded webhook event.

As far as I know, the is no way to check if the card is issued by an Indian bank. Although, now that you mention it, we can consider passing them only for orders using INR currency. Stripe will only return a mandate ID for those cases that require it.

>Is mandate tied to specific card? If someone knows a mandate id, can they use it to create a payment intent with different card and by pass the needed customer confirmation?

The mandate is tied to both the card and subscription. Using an existing mandate ID in an intent with a different card it will fail and show the `Not matching mandate` error shown on the PR description.

>Instruction to test WCPay Subscription renewal can be found here: PCYsg-DQ4-p2.

Thanks! I'll test it and update the PR accordingly soon.

>I saw this PR branch was pretty behind develop, so I merged develop into fix/2935-recurring-transactions-indian-ccs. I hope it's OK.

Sure! 🙂 

> Will this be addressed, maybe in different PR? If so, please create a separate issue so it is not forgotten.
>
>Changing payment method is part of the critical flows ([ref](https://github.com/Automattic/woocommerce-payments/wiki/WC-Pay-Subscriptions-critical-flows-testing-instructions#change-payment-method-card-for-an-active-subscription)) but I bet it won't be detected as bug because GlobalStep likely will test using US card.

Yes, I plan to create a followup issue for both of the known issues before merging this. The PR has been blocked for more than a year, and delaying it again to fix those issues could end up in another big refactor due to planned changes in the affected code. I think it's safe to proceed with it, as it will at least allow renewing subscriptions without errors.

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (shendy-a8c) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:900]: That makes sense! No need to check then.

File: includes/class-wc-payments-webhook-processing-service.php
Comment (shendy-a8c) [reply includes/class-wc-payments-webhook-processing-service.php:366]: Thank you for explaining all that. Good to know that `data.object.last_payment_error` will always be available and it's a better alternative than reading from charges.

File: includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php
Comment (shendy-a8c) [reply includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php:900]: As long as intended and not an oversight, it's good.

**Comment by shendy-a8c** [comment]
> As far as I know, the is no way to check if the card is issued by an Indian bank. Although, now that you mention it, we can consider passing them only for orders using INR currency. Stripe will only return a mandate ID for those cases that require it.

Cool. I was just checking. I don't think there is any change needed here.

> Although, now that you mention it, we can consider passing them only for orders using INR currency.

I think checking for currency is not ideal because the requirement is on the card itself, ie if it's issued by India bank, despite the currency, right?

**Comment by ismaeldcom** [comment]
>I think checking for currency is not ideal because the requirement is on the card itself, ie if it's issued by India bank, despite the currency, right?

Yep, Stripe mentions `Currently, all issuing banks only support INR for subscriptions.` but that could change in the future. So better to keep it as is.

**Comment by ismaeldcom** [comment]
Hey @shendy-a8c! Now that every comment is resolved, ~~I'm going to create the follow-up issues~~ done, and proceed with the merge. Anything else on your side before proceeding?

#### Follow-up issues:
- [Improve recurring payments order notes on Indian CCs](https://github.com/automattic/woocommerce-payments/issues/3075#top)
- [Handle changing payment method for orders with a mandate](https://github.com/Automattic/woocommerce-payments/issues/5249#top)

After testing early renewals using WCPay Subscriptions it seems to work as expected. But as it relies on Stripe Billing the mandate is not attached to the intent. So I'm thinking to create another issue to confirm, by asking Stripe first, if we also need to do something in our end for [India recurring payments](https://stripe.com/docs/india-recurring-payments) using the Billing API.

If you agree with the mentioned idea, would you mind creating the WCPay Subscription issue? I'm sure you have way more knowledge about how it works than me, and could add a better description 🙂

**Review by shendy-a8c — APPROVED** [review]

**Comment by shendy-a8c** [comment]
Sounds like a plan, @ismaeldcom.

> After testing early renewals using WCPay Subscriptions it seems to work as expected.

I think 'early renewal' is not the term you meant here. WCPay Subscriptions does not support early renewals ([ref](https://woocommerce.com/document/payments/subscriptions-basics/comparison/#feature-comparison)). With WooCommerce Subscriptions, early renewal is possible. Early renewal means one (customer or merchant) can renew the subscriptions before the expected renewal date. Testing WCPay Subscriptions with billing clock even though feels like renewing the subscriptions before the expected renewal date, it is not an early renewal because it is just a way to move the time forward artificially.

> But as it relies on Stripe Billing the mandate is not attached to the intent.

Any link to the payment intent you tested? It might be because you tested with billing clock.

> would you mind creating the WCPay Subscription issue?

I like to make sure it's really a problem by testing without billing clock. More details can be found here: pdjTHR-21b-p2.

As of this PR, I think it looks good and I think it's a good idea to wrap up and follow up with new issues you created and one that I might create after I can confirm there is a problem with WCPay Subscriptions.

**Comment by ismaeldcom** [comment]
>I think 'early renewal' is not the term you meant here

You're right! I didn't choose the best name here 🙂  I meant what you explained, using the test billing clock.

>Any link to the payment intent you tested? It might be because you tested with billing clock.

Sure, here is one example: https://dashboard.stripe.com/test/connect/accounts/acct_1Lwn9BQrnu7Ppo4o/payments/pi_3MByt5Qrnu7Ppo4o1TeYLfEs

And yes, I think that's the cause. If I understood the way it works, it creates a test subscription (with a test customer) under the hood. So I think it's expected to not attach the mandate.

Thanks approving the PR and organizing the P2 with further testing for WCPay Subscriptions 👍

**Comment by shendy-a8c** [comment]
I just created https://github.com/Automattic/woocommerce-payments/issues/5311 to fix WCPay Subscriptions renewals with Indian CC.

**Pull request merged**
